src/Type/EntityStorage/EntityStorageDynamicReturnTypeExtension.php

```diff
@@ -60,6 +61,18 @@ public function getTypeFromMethodCall(
         $callerType = $scope->getType($methodCall->var);
 
         if (!$callerType instanceof EntityStorageType) {
+            if (!$callerType instanceof ObjectType) {
+                return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+            }
+
+            // Workaround because cannot figure out why the caller type is not an EntityStorageType
+            // when it has been type hinted.
+            // Instead, we try to infer the type, i.e. ContentEntityStorageType or ConfigEntityStorageType, here.
+            // @todo: we should definitively look for other cases that getQuery.
+            if ($methodReflection->getName() === 'getQuery') {
+                return $this->getReturnTypeForGetQueryMethod($methodReflection, EntityStorageHelper::getTypeFromStorageObjectType($callerType));
```

Comment:
If we really don't find another way to fix the issue, instead of this we should refactor the code to do something like:

```php
$callerType = $scope->getType($methodCall->var);

if (!$callerType instanceof EntityStorageType && !$callerType instanceof ObjectType) {
    return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
}

$entityStorageType = $callerType instanceof EntityStorageType ? $callerType : EntityStorageHelper::getTypeFromStorageObjectType($callerType);

// refactor the rest to use $entityStorageType instead of $callerType
```